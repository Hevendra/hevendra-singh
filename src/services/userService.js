export const userService = {
  fetchUsers
};

function fetchUsers(pageNo, limit) {
  return fetch(
    "https://reqres.in/api/users?page=" + pageNo + "&per_page=" + limit
  )
    .then(response => {
      if (!response.ok) {
        return Promise.reject(response.statusText);
      }
      return response.json();
    })
    .then(user => {
      return user;
    })
    .catch(e => {
      alert("Something went wrong!!");
    });
}
