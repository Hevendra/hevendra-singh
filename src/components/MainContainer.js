import React, { Component } from "react";
import { userService } from "../services";
import UserDetail from "./userDetailComponent";
import Pagination from "./paginationComponent";
import "../App.css";

export default class MainContainer extends Component {
  constructor() {
    super();
    this.state = { userList: [], pageNo: 0, isAllFetched: false };
  }

  componentDidMount() {
    this.fetchUserFromApi(this.state.pageNo + 1, 10);
  }

  fetchUserFromApi(pageNo, numberOfRecords) {
    userService.fetchUsers(pageNo, numberOfRecords).then(
      data => {
        this.setState({
          userList: data.data,
          pageNo: data.page,
          isAllFetched: data.page === data.total_pages
        });
      },
      error => {
        console.log(error);
      }
    );
  }

  deleteUserById(e, userId) {
    let updatedUsersList = this.state.userList.filter(
      data => data.id != userId
    );
    this.setState({ userList: updatedUsersList });
  }

  deleteDuplicateUsers(e) {
    const updatedUsersList = this.state.userList.reduce(
      (x, y) =>
        x.findIndex(e => e.first_name === y.first_name) < 0 ? [...x, y] : x,
      []
    );
    this.setState({ userList: updatedUsersList });
    e.preventDefault();
  }

  showUsers() {
    let users = [];
    if (this.state.userList.length) {
      this.state.userList.map((user, i) => {
        users.push(
          <UserDetail
            key={i}
            data={user}
            handleDelete={e => this.deleteUserById(e, user.id)}
          />
        );
      });
    } else {
      users.push(
        <div key={1} className="display-name">
          No User data found!
        </div>
      );
    }
    return users;
  }

  render() {
    return (
      <div className="App">
        <div className="header">
          <h1 className="App-title">Pearson User Management</h1>
          <Pagination
            pageNo={this.state.pageNo}
            isAllFetched={this.state.isAllFetched}
            fetchUserFromApi={this.fetchUserFromApi.bind(this)}
          />
        </div>
        <div className="Container">{this.showUsers()}</div>
        {this.state.userList.length ? (
          <div className="delete-duplicate">
            <a href="" onClick={e => this.deleteDuplicateUsers(e)}>
              Delete Duplicate
            </a>
          </div>
        ) : (
          ""
        )}
      </div>
    );
  }
}
