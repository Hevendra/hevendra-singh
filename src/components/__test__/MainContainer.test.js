import React from "react";
import { mount } from "enzyme";
import MainContainer from "../MainContainer";
import Pagination from "../paginationComponent";
import { userService } from "../../services";

describe("<MainContainer/>", () => {
  const sampleTestData = {
    page: 2,
    per_page: 10,
    total: 12,
    total_pages: 2,
    data: [
      {
        id: 11,
        email: "george.edwards@reqres.in",
        first_name: "George",
        last_name: "Edwards",
        avatar:
          "https://s3.amazonaws.com/uifaces/faces/twitter/mrmoiree/128.jpg"
      },
      {
        id: 12,
        email: "rachel.howell@reqres.in",
        first_name: "Rachel",
        last_name: "Howell",
        avatar:
          "https://s3.amazonaws.com/uifaces/faces/twitter/hebertialmeida/128.jpg"
      }
    ]
  };
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<MainContainer />);
  });

  afterEach(() => {
    wrapper.unmount();
  });
  it("should render MainContainer correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render Paggination Component", () => {
    expect(wrapper.containsMatchingElement(<Pagination />)).toEqual(true);
  });

  it("should delete user profile", () => {
    const instance = wrapper.instance();
    wrapper.setState({
      userList: sampleTestData.data
    });
    instance.deleteUserById(0, 11);
    expect(wrapper.state("userList").length).toEqual(1);
  });

  it("should show currect number of user profiles", () => {
    wrapper.setState({
      userList: sampleTestData.data
    });
    expect(wrapper.find(".card-body").length).toEqual(2);
  });
});

describe("MainContainer> Testing User Service", () => {
  const sampleTestData = {
    page: 2,
    per_page: 10,
    total: 12,
    total_pages: 2,
    data: [
      {
        id: 11,
        email: "george.edwards@reqres.in",
        first_name: "George",
        last_name: "Edwards",
        avatar:
          "https://s3.amazonaws.com/uifaces/faces/twitter/mrmoiree/128.jpg"
      },
      {
        id: 12,
        email: "rachel.howell@reqres.in",
        first_name: "Rachel",
        last_name: "Howell",
        avatar:
          "https://s3.amazonaws.com/uifaces/faces/twitter/hebertialmeida/128.jpg"
      }
    ]
  };

  beforeEach(() => {
    fetch.resetMocks();
  });
  window.alert = jest.fn();
  it("should user-api retun correct number of records", () => {
    fetch.mockResponseOnce(JSON.stringify(sampleTestData));

    //assert on the response
    userService.fetchUsers(2, 2).then(data => {
      expect(data.data.length).toEqual(2);
    });

    //assert on the times called and arguments given to fetch
    expect(fetch.mock.calls[0][0]).toEqual(
      "https://reqres.in/api/users?page=2&per_page=2"
    );
  });
});
