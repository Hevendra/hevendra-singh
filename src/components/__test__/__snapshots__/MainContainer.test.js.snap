// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainContainer/> should render MainContainer correctly 1`] = `
<MainContainer>
  <div
    className="App"
  >
    <div
      className="header"
    >
      <h1
        className="App-title"
      >
        Pearson User Management
      </h1>
      <Pagination
        fetchUserFromApi={[Function]}
        isAllFetched={false}
        pageNo={0}
      >
        <div
          className="pagination"
        >
          <a
            className="disabled"
            href="#"
            onClick={[Function]}
          >
            ❮❮
          </a>
          <a
            className="active"
            href="#"
            onClick={[Function]}
          >
            ❯❯
          </a>
        </div>
      </Pagination>
    </div>
    <div
      className="Container"
    >
      <div
        className="display-name"
        key="1"
      >
        No User data found!
      </div>
    </div>
  </div>
</MainContainer>
`;
